sockeye/layers.py

```diff
@@ -46,11 +48,1542 @@ def activation(data: mx.sym.Symbol, act_type: str) -> mx.sym.Symbol:
+            return self.num_hidden
+
+    def _ff(self, data: mx.sym.Symbol):
+        h = mx.sym.FullyConnected(data=data, num_hidden=self._pre_activation_num_hidden(),
+                                  weight=self.w_i2h, bias=self.b_i2h,
+                                  flatten=False, name=self.prefix + "_ff")
+
+        if self.act_type == C.GLU:
+            # GLU
+            # two times: (batch_size, seq_len, num_hidden)
+
+            # pylint: disable=unbalanced-tuple-unpacking
+            gate_a, gate_b = mx.sym.split(h, num_outputs=2, axis=2)
+            # (batch_size, seq_len, num_hidden)
+            h = mx.sym.broadcast_mul(gate_a,
+                                     mx.sym.Activation(data=gate_b, act_type="sigmoid"))
+        else:
+            h = activation(h, act_type=self.act_type)
+        if self.dropout > 0.0:
+            h = mx.sym.Dropout(h, p=self.dropout)
+        return h
+
+    def __call__(self,
+                 data: mx.sym.Symbol,
+                 lengths: Optional[mx.sym.Symbol] = None,
+                 max_length: Optional[int] = None) -> mx.sym.Symbol:
+        """
+        Position-wise feed-forward network with activation.
+
+        :param x: Symbol of shape (batch_size, seq_len, num_hidden)
+        :return: Symbol of shape (batch_size, seq_len, num_hidden)
+        """
+        return self._ff(data)
+
+    def get_num_hidden(self) -> int:
+        return self.num_hidden
+
+
+class FeedForwardLayerConfig(LayerConfig):
+
+    def __init__(self,
+                 num_hidden: int,
+                 dropout: float,
+                 act_type: str = C.RELU):
+        super().__init__()
+        self.num_hidden = num_hidden
+        self.act_type = act_type
+        self.dropout = dropout
+
+    def create_encoder_layer(self, input_num_hidden: int, prefix: str) -> EncoderLayer:
+        return self.create_block_layer(prefix)
+
+    def create_decoder_layer(self, input_num_hidden: int, prefix: str) -> DecoderLayer:
+        return self.create_block_layer(prefix)
+
+    def create_block_layer(self, prefix: str) -> StatelessBlockLayer:
+        return StatelessBlockLayer(FeedForwardBlock(self.num_hidden,
+                                                    self.dropout,
+                                                    self.act_type,
+                                                    prefix + "ff_"))
+
+
+class LinearBlock(StatelessBlock):
+    """
+    Position-wise feed-forward network with activation.
+    """
+
+    def __init__(self,
+                 num_hidden: int,
+                 dropout: float,
+                 no_bias: bool,
+                 prefix: str) -> None:
+        self.num_hidden = num_hidden
+        self.dropout = dropout
+        self.no_bias = no_bias
+        self.prefix = prefix
+        self.w_i2h = mx.sym.Variable('%si2h_weight' % prefix)
+        self.b_i2h = None if no_bias else mx.sym.Variable('%si2h_bias' % prefix)
+
+    def _fc(self, data: mx.sym.Symbol):
```

Comment:
why factoring this into a private function? is this used elsewhere then __call__?

Reply:
that's an artifact of an earlier version where a `StatelessBlock`  always had to implement both `__call__` and `step`. I'll move everything back to  `__call__` (also for FF).